### PR TITLE
LwM2m JSON format impl

### PR DESCRIPTION
**Pull request by alseddnm**
This is our 1st attempt to make a major contribution in Leshan, let us know if our code does not meet your conventions and standards, we are very flexible to make any modifications.

### CODE REVIEW

File: leshan-core/pom.xml [hunk 1]
```diff
@@ -38,7 +38,10 @@ Contributors:
             <groupId>com.google.code.gson</groupId>
             <artifactId>gson</artifactId>
         </dependency>
-
+		<dependency>
```
Comment (jvermillard) [inline leshan-core/pom.xml]: formating ? ;(

File: leshan-core/src/main/java/org/eclipse/leshan/core/model/DDFFileParser.java [hunk 2]
```diff
@@ -169,6 +169,10 @@ private ResourceModel parseResource(Node item) {
                 case "Time":
                     type = Type.TIME;
                     break;
+                case "Objlnk":
```
Comment (jvermillard) [inline leshan-core/src/main/java/org/eclipse/leshan/core/model/DDFFileParser.java]: IMO adding Object link should be part of another pull request, it's a different feature

File: leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java [hunk 3]
```diff
@@ -109,6 +122,13 @@ public static LwM2mNode decode(byte[] content, ContentFormat format, LwM2mPath p
             }
             return new LwM2mResource(path.getResourceId(), Value.newBinaryValue(content));
         case JSON:
+        	try {
```
Comment (jvermillard) [inline leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java]: formating?

File: leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java [hunk 4]
```diff
@@ -109,6 +122,13 @@ public static LwM2mNode decode(byte[] content, ContentFormat format, LwM2mPath p
...
+        		String jsonStrValue = new String(content);
+				LwM2mJsonObject json =  LwM2mJson.fromJsonLwM2m(jsonStrValue);
+				return parseJSON(json,path, model);
+			} catch (LwM2mJsonException e) {
+			    throw new InvalidValueException("Unable to deSerialize json.", path, e);
```
Comment (jvermillard) [inline leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java]: remove '.'  at the end of the exception message

File: leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java [hunk 5]
```diff
@@ -210,7 +247,83 @@ private static LwM2mNode parseTlv(Tlv[] tlvs, LwM2mPath path, LwM2mModel model)
...
     }
+    
+    private static LwM2mNode parseJSON(LwM2mJsonObject jsonObject, LwM2mPath path, LwM2mModel model) throws InvalidValueException {
+        LOG.trace("Parsing JSON content for path {}: {}", path, jsonObject);
+
+        if (path.isObject()) {
+           // TODO 
+           // If bn is present will have multiple object instances in JSON the payload
+           // In case of JSON contains ObjLnk -> this should return List<LwM2mNode> ???
+        	 throw new InvalidValueException("Not yet implemented: " + " Object Path ", path);
```
Comment (jvermillard) [inline leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java]: throw new InvalidStateException("Not implemented")

File: leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeEncoder.java [hunk 6]
```diff
@@ -85,7 +90,13 @@
...
             break;
         case JSON:
-            throw new IllegalArgumentException("JSON content format not supported");
+        	NodeJsonEncoder jsonEncoder = new NodeJsonEncoder();
+        	jsonEncoder.objectId = path.getObjectId();
+        	jsonEncoder.model = model;
+            node.accept(jsonEncoder);
+            encoded = jsonEncoder.encoded;
+          break;
+//            throw new IllegalArgumentException("JSON content format not supported");
```
Comment (jvermillard) [inline leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeEncoder.java]: remove this line

File: leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+import java.io.Serializable;
+
+import com.google.gson.annotations.SerializedName;
+
+public class JsonArrayElement implements Serializable {
+
+	/**
+	 * 
+	 */
+	private static final long serialVersionUID = -7863478129361403779L;
```
Comment (jvermillard) [inline leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java]: please remove , if you really want it use version 1

File: leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+				return false;
+		} else if (!stringValue.equals(other.stringValue))
+			return false;
+		if (time != other.time)
+			return false;
+		return true;
+	}
+
+	@Override
+	public String toString() {
```
Comment (jvermillard) [inline leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java]: StringFormat please

File: leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJson.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+public class LwM2mJson {
+	
+	private static final Logger LOG = LoggerFactory.getLogger(LwM2mJson.class);
+			
+	private static final Gson gson = new GsonBuilder().create();
+	
+	
+    /**
+     * <ul> This <code>toJson</code> method accepts LwM2mJsonObject object and converts to JSON String.
+	 * <li>
```
Comment (jvermillard) [inline leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJson.java]: ?

File: leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJson.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    */
+	public static LwM2mJsonObject fromJsonLwM2m(String jsonString)
+			throws LwM2mJsonException {
+		try {
+			LwM2mJsonObject element = gson.fromJson(jsonString,
+					LwM2mJsonObject.class);
+			return element;
+		} catch (Exception e) {
+			LOG.error(" Unable to deserialize JSON String to LwM2mJsonElement object: " +e.getMessage());
+			throw new LwM2mJsonException(
```
Comment (jvermillard) [inline leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJson.java]: log or throw, you need to choose :)

File: leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJsonObject.java [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+package org.eclipse.leshan.json;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+
+import com.google.gson.annotations.SerializedName;
+
+public class LwM2mJsonObject implements Serializable{
+	
+	private static final long serialVersionUID = 7302201912160751832L;
```
Comment (jvermillard) [inline leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJsonObject.java]: please use version 1

File: leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJsonObject.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+			if (other.jsonArray != null)
+				return false;
+		} else if (!jsonArray.equals(other.jsonArray))
+			return false;
+		return true;
+	}
+
+	@Override
+	public String toString() {
+		return "LwM2mJsonElement [baseName=" + baseName + ", baseTime="
```
Comment (jvermillard) [inline leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJsonObject.java]: StringFormat

File: leshan-core/src/test/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoderTest.java [hunk 13]
```diff
@@ -152,4 +167,53 @@ public void tlv_power_source__multiple_resource() throws InvalidValueException {
         assertEquals(1, ((Number) resource.getValues()[0].value).intValue());
         assertEquals(5, ((Number) resource.getValues()[1].value).intValue());
     }
+    
+    @Test
+    public void json_device_object_instance0() throws InvalidValueException {
+        // tlv content for instance 0 of device object
+    	String content = "{"
+    			+ "\"e\":["+"{"+"\"n\":\"0\","+"\"sv\":\"Open Mobile Alliance\""+"},"
```
Comment (jvermillard) [inline leshan-core/src/test/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoderTest.java]: please remove the unused "+"

File: leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java (hunk 5 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java]: Where to find InvalidStateException?
I don't think is part of the JDK, do you mean IllegalStateException?

Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java]: I can also Implement InvalidStateException.java, But it will be used only for this purpose?

File: leshan-core/src/main/java/org/eclipse/leshan/core/model/DDFFileParser.java (hunk 2 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/core/model/DDFFileParser.java]: done.

File: leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java (hunk 3 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java]: done.

File: leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java (hunk 4 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java]: done.

File: leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeEncoder.java (hunk 6 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeEncoder.java]: done.

File: leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java (hunk 7 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java]: done.

File: leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJson.java (hunk 9 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJson.java]: done.

File: leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJson.java (hunk 10 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJson.java]: done.

File: leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJsonObject.java (hunk 11 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJsonObject.java]: done.

File: leshan-core/src/test/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoderTest.java (hunk 13 above)
Comment (alseddnm) [inline leshan-core/src/test/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoderTest.java]: done.

File: leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJsonObject.java (hunk 12 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJsonObject.java]: done.

File: leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java (hunk 8 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java]: You mean to use String.format() instead of StringBuilder?

File: leshan-core/pom.xml (hunk 1 above)
Comment (alseddnm) [inline leshan-core/pom.xml]: done.

File: leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java (hunk 5 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java]: InvalidStateException not part of the JDK and Leshan don't have it. Do you mean to implement it?

Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoder.java]: @jvermillard ???

File: leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java (hunk 8 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java]: @jvermillard ???

**Comment by alseddnm** [comment]
@jvermillard We have all the changes in our fork JSON format branch, how do we create a new pull request. Should we close this one 1st?
Sorry we kind of learning GitHub process.
Thank you,
Nizar...!

**Comment by sophokles73** [comment]
Hi Nizar,

you do not necessarily create a new pull request since GitHub automatically adds your subsequent changes to the existing PR. However, some of the authors do not seem to have a valid CLA in place and in some commits the SIgned-off-by footer does not match the author of the commit :-(
From taking a look at the [issues GitHub complained about](https://dev.eclipse.org/eclipse-webhook/services/status_details.php?id=5554cde6d1335) it looks like you used two different email addresses to log into GitHub ...

Kai

**Comment by alseddnm** [comment]
@sophokles73 I have corrected the email address issue, changed it in my GitHub and Eclipse account to use my gmail. As of this pull request should I do any action or just leave it as is?

File: leshan-core/src/main/java/org/eclipse/leshan/core/model/DDFFileParser.java [hunk 14]
```diff
@@ -169,7 +169,7 @@ private ResourceModel parseResource(Node item) {
                 case "Time":
                     type = Type.TIME;
                     break;
-                }
+                 }
```
Comment (jvermillard) [inline leshan-core/src/main/java/org/eclipse/leshan/core/model/DDFFileParser.java]: WTF?

File: leshan-core/pom.xml [hunk 15]
```diff
@@ -1,93 +1,86 @@
-<!--
```
Comment (jvermillard) [inline leshan-core/pom.xml]: Why you reformatted this file?

File: leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java (hunk 8 above)
Comment (jvermillard) [inline leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java]: look at the other toString in the Leshan code, String.format 
look at https://github.com/jvermillard/leshan/wiki/Code-%26-design-guidelines

File: leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJsonException.java [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+	public LwM2mJsonException(String message) {
+        super(message);
+    }
+
+    public LwM2mJsonException(String message, Exception cause) {
+        super(message, cause);
+    }
+
+    private static final long serialVersionUID = 5765213465835546341L;
```
Comment (jvermillard) [inline leshan-core/src/main/java/org/eclipse/leshan/json/LwM2mJsonException.java]: version 1

File: leshan-core/src/test/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoderTest.java [hunk 17]
```diff
@@ -15,7 +15,9 @@
  *******************************************************************************/
 package org.eclipse.leshan.core.node.codec;
 
-import static org.junit.Assert.*;
```
Comment (jvermillard) [inline leshan-core/src/test/java/org/eclipse/leshan/core/node/codec/LwM2mNodeDecoderTest.java]: keep the static import at it was

File: leshan-core/src/main/java/org/eclipse/leshan/core/model/DDFFileParser.java (hunk 14 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/core/model/DDFFileParser.java]: @jvermillard  what you mean by WTF?
We removed the OBJECTLINK, as you said it has to be in a different PR.

File: leshan-core/pom.xml (hunk 15 above)
Comment (alseddnm) [inline leshan-core/pom.xml]: sorry I will revert this change..

File: leshan-core/src/main/java/org/eclipse/leshan/core/model/DDFFileParser.java (hunk 14 above)
Comment (jvermillard) [inline leshan-core/src/main/java/org/eclipse/leshan/core/model/DDFFileParser.java]: you just added a space char to this file

File: leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java (hunk 8 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/json/JsonArrayElement.java]: @jvermillard  Yes I did read the guidelines, However toString has conditional statements in this case, I don't think is good to have an empty string printed in toString output?
The value could be float, String, Boolean, string or objectLink. Only one value will be displayed in the to Sting output. The time value also can be empty.
I have seen StringBuilder has been used  in Leshan Core (ResourceModel.java, ObjectModel.java,...) ..
Let me know what you think?

Thanks

File: leshan-core/src/main/java/org/eclipse/leshan/core/model/DDFFileParser.java (hunk 14 above)
Comment (alseddnm) [inline leshan-core/src/main/java/org/eclipse/leshan/core/model/DDFFileParser.java]: removed ..

**Comment by alseddnm** [comment]
@jvermillard  This looks OK to you?
Thanks,
Nizar..!

**Comment by jvermillard** [comment]
Can you fix this? : 

    The following users do not have valid CLA. Please see this page for details, or to sign a CLA.
        nalsedde@gemalto.com
        nalsedde@dvn8bs1.gemalto.com

    The following users have invalid Signed-off-by footers. Please use the Signed-Off-By (git commit -s) mechanism to confirm the origin of the submission. The identity in the Signed-off-By must match the Author or the Committer.
        nalsedde@gemalto.com
        nalsedde@dvn8bs1.gemalto.com
